src/manager.c

```diff
@@ -162,48 +178,64 @@ register_new_comp(char *filename, bool allow_default_entry)
 struct CompMapping *
 mapping_new(struct Compartment *to_map)
 {
+    if (!mappings)
+    {
+        mappings = mappings_init();
+    }
+
     return mapping_new_fixed(to_map, NULL);
 }
 
 struct CompMapping *
 mapping_new_fixed(struct Compartment *to_map, void *addr)
 {
-    int mmap_flags = MAP_PRIVATE | MAP_ANONYMOUS;
-    if (addr != NULL)
+    struct CompMapping *cached_mapping;
+    cached_mapping = mappings_search_free(to_map, mappings);
+    if (cached_mapping)
     {
-        assert((uintptr_t) addr % to_map->page_size == 0);
-        mmap_flags |= MAP_FIXED;
+        mapping_reuse(cached_mapping);
+        return cached_mapping;
     }
-    // Map new compartment
-    void *map_result = mmap(addr, to_map->total_size,
-        PROT_READ | PROT_WRITE | PROT_EXEC, mmap_flags, -1, 0);
-    if (map_result == MAP_FAILED)
+    else
     {
-        err(1, "Error mapping compartment %zu data at addr %p", to_map->id,
-            addr);
+        int mmap_flags = MAP_PRIVATE | MAP_ANONYMOUS;
+        if (addr != NULL)
+        {
+            assert((uintptr_t) addr % to_map->page_size == 0);
+            mmap_flags |= MAP_FIXED;
+        }
+        // Map new compartment
+        void *map_result = mmap(addr, to_map->total_size,
+            PROT_READ | PROT_WRITE | PROT_EXEC, mmap_flags, -1, 0);
+        if (map_result == MAP_FAILED)
+        {
+            err(1, "Error mapping compartment %zu data at addr %p", to_map->id,
+                addr);
+        }
+        addr = map_result;
     }
-    addr = map_result;
 
-    memcpy(addr, to_map->staged_addr, to_map->total_size);
+    memcpy(addr, to_map->staged_addr,
+        to_map->data_size + to_map->scratch_mem_extra);
 
     // Set appropriate `mprotect` flags
-    struct LibDependency *lib_dep;
-    struct SegmentMap lib_dep_seg;
-    for (size_t i = 0; i < to_map->libs_count; ++i)
-    {
-        lib_dep = to_map->libs[i];
-        for (size_t j = 0; j < lib_dep->lib_segs_count; ++j)
-        {
-            lib_dep_seg = lib_dep->lib_segs[j];
-            if (mprotect(get_seg_target(addr, lib_dep, j), lib_dep_seg.mem_sz,
-                    lib_dep_seg.prot_flags)
-                != 0)
-            {
-                err(1, "Error setting permissions for %p (lib %zu seg %zu)",
-                    get_seg_target(addr, lib_dep, j), i, j);
-            }
-        }
-    }
+    /*struct LibDependency *lib_dep;*/
```

Comment:
Why is this commented?

Reply:
We had some performance concerns about this, so I initially commented it out for benchmarking. However, with the new reuse implementation, this should be short-circuited earlier on, so it will likely not be a significant performance hit in future benchmarking. Uncommented in https://github.com/capablevms/CHERI-ELF-comp/pull/44/commits/6277ba919eb49e7e1af831077f857629e1ea6038.

Reply:
Yes, sorry, both of them together fix this issue. I am just extremely frustrated with the version mismatch of `clang-format` between my local device and CHERI and all the weird `ssh`ing I had to do...